rust/rust.cc

```diff
@@ -101,31 +100,31 @@ extern "C" {
     Loc npnr_context_get_pip_location(const Context *ctx, uint64_t pip) { return ctx->getPipLocation(unwrap_pip(pip)); }
     bool npnr_context_check_pip_avail_for_net(const Context *ctx, uint64_t pip, NetInfo *net) { return ctx->checkPipAvailForNet(unwrap_pip(pip), net); }
 
-    uint64_t npnr_context_get_pips_leak(const Context *ctx, uint64_t **pips) {
-        auto size = size_t{};
-        for (auto _pip : ctx->getPips()) {
-            NPNR_UNUSED(_pip);
-            size++;
-        }
+    uint64_t npnr_context_get_pips_leak(const Context *const ctx, uint64_t **const pips) {
+        const auto ctx_pips{ctx->getPips()};
+        const auto size{
+            std::accumulate(ctx_pips.begin(), ctx_pips.end(), /*initial value*/ size_t{},
+                [](size_t value, const auto &/*pip*/) { return value + 1U; }
+        )};
         *pips = new uint64_t[size];
         auto idx = 0;
-        for (auto pip : ctx->getPips()) {
+        for (const auto &pip : ctx_pips) {
             (*pips)[idx] = wrap(pip);
             idx++;
         }
         // Yes, by never deleting pip_vec, we leak memory.
         return size;
     }
 
-    uint64_t npnr_context_get_wires_leak(const Context *ctx, uint64_t **wires) {
-        auto size = size_t{};
-        for (auto _wire : ctx->getWires()) {
-            NPNR_UNUSED(_wire);
-            size++;
-        }
+    uint64_t npnr_context_get_wires_leak(const Context *const ctx, uint64_t **const wires) {
+        const auto ctx_wires{ctx->getWires()};
+        const auto size{
+            std::accumulate(ctx_wires.begin(), ctx_wires.end(), /*initial value*/ size_t{},
+                [](size_t value, const auto &/*wire*/) { return value + 1U; }
+        )};
         *wires = new uint64_t[size];
         auto idx = 0;
-        for (auto wire : ctx->getWires()) {
+        for (const auto &wire : ctx_wires) {
```

Comment:
there isn't a need for making this a reference, as `WireId` is not expected (and indeed, required not to be in this case), more than a 64 bit value

Reply:
If this is done without being `const auto &`, you will always copy and as this is not a trivial type (which is why the warnings btw) that will always generate a call to the copy constructor, even though that might be able to fit in a 64-bit register in your mental model. This is expensive and avoidable - similarly, it will generate destructor calls after, potentially invalidating the wrapped object state that's just been copied.

This construction guarantees this operation to remain cheap, no matter how `WireId` winds up implemented in the future, and the wrapped objects to remain valid post-loop.

Reply:
Everything in nextpnr would break, at a minimum from a performance PoV, if the ID types had non-trivial copies or destructors - the entire principle throughout nextpnr to pass them by value.

Reply:
One of the backends makes WireId a IdString, another it's CycloneV::rnode_t, etc (couldn't reason about triviality).. there are several where it's two `int32_t`'s, which will spill to two registers here. Using `const auto &` prevents this and any possible impact and eliminates a copy. Using a `const auto &` here also ensures that we only copy once - in wrap() - from the storage in the holder type directly into the FFI boundary storage.

Iff WireId was fully POD in all cases, GCC would not have warned. It's telling us that the type is non-POD (so also non-trivial) even if it is cheap-ish to copy.

Reply:
`rnode_t` is just a typedef to an `int32_t` and an `IdString` is a class with a single `int32_t` member. 

I think most of this "cheap-ish" copies should be optimised away, but hopefully so will the reference here, so fine, I won't bother arguing further.

Reply:
> there are several where it's two int32_t's, which will spill to two registers here

That doesn't make the type non-POD. I see you use the same argument strategy here as you did on the Amaranth channels. It is equally unwelcome.